pyomo/contrib/cp/interval_var.py

```diff
@@ -0,0 +1,202 @@
+from pyomo.core.base.global_set import UnindexedComponent_index
+from pyomo.core.base.indexed_component import (
+    IndexedComponent, UnindexedComponent_set)
+from pyomo.core.base.initializer import BoundInitializer, Initializer
+from pyomo.core.expr.current import GetAttrExpression, GetItemExpression
+
+class IntervalVarTimePoint(ScalarVar):
+    """This class defines the abstract interface for a single variable
+    denoting a start or end time point of an IntervalVar"""
+
+    __slots__ = ()
+
+    def get_associated_interval_var(self):
+        return self.parent_block()
+
+    def before(self, time, delay=0):
+        return BeforeExpression((self, time, delay))
+
+    def after(self, time, delay=0):
+        return BeforeExpression((time, self, delay))
+
+    def at(self, time, delay=0):
+        return AtExpression((self, time, delay))
+
+
+class IntervalVarStartTime(IntervalVarTimePoint):
+    """This class defines a single variable denoting a start time point
+    of an IntervalVar"""
+    def __init__(self):
+        super().__init__(domain=Integers, ctype=IntervalVarStartTime)
+
+
+class IntervalVarEndTime(IntervalVarTimePoint):
+    """This class defines a single variable denoting an end time point
+    of an IntervalVar"""
+    def __init__(self):
+        super().__init__(domain=Integers, ctype=IntervalVarEndTime)
+
+
+class IntervalVarLength(ScalarVar):
+    """This class defines the abstract interface for a single variable
+    denoting the length of an IntervalVar"""
+
+    __slots__ = ()
+
+    def __init__(self):
+        super().__init__(domain=Integers, ctype=IntervalVarLength)
+
+    def get_associated_interval_var(self):
+        return self.parent_block()
+
+
+class IntervalVarPresence(ScalarBooleanVar):
+    """This class defines the abstract interface for a single Boolean variable
+    denoting whether or not an IntervalVar is scheduled"""
+
+    __slots__ = ()
+
+    def __init__(self):
+        super().__init__(ctype = IntervalVarPresence)
+
+    def get_associated_interval_var(self):
+        return self.parent_block()
+
+
+class IntervalVarData(_BlockData):
+    """This class defines the abstract interface for a single interval variable.
+    """
+    # We will put our four variables on this, and everything else is off limits.
+    _Block_reserved_words = Any
+
+    def __init__(self, component=None):
+        _BlockData.__init__(self, component)
+
+        with self._declare_reserved_components():
+            self.is_present = IntervalVarPresence()
+            self.start_time = IntervalVarStartTime()
+            self.end_time = IntervalVarEndTime()
+            self.length = IntervalVarLength()
+
+    @property
+    def optional(self):
+        # We only store this information in one place, but it's kind of annoying
+        # to have to check if the BooleanVar is fixed, so this way you can ask
+        # the IntervalVar directly.
+        return not self.is_present.fixed
```

Comment:
Is optional really `False` if `is_present.fix(False)`?

Reply:
Eeeek, no, good catch. It had better be optional in that case.